domain/application/state/application.go

```diff
@@ -1179,6 +1180,217 @@ WHERE a.name = $applicationName.name
+				return errors.Capture(err)
+			}
+			endpointBindings, err = st.getEndpointBindings(ctx, tx, appUUID)
+			if err != nil {
+				return errors.Capture(err)
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, errors.Errorf("querying application %q addresses hash: %w", appUUID, err)
+		}
+		hash, err := st.hashAddressesAndEndpoints(spaceAddresses, endpointBindings)
+		if err != nil {
+			return nil, errors.Capture(err)
+		}
+		return []string{hash}, nil
+	}
+	return "ip_address", "application_endpoint", queryFunc
+}
+
+// GetNetNodeUUIDByUnitName returns the net node UUID for the named unit or the
+// cloud service associated with the unit's application. This method is meant
+// to be used in the WatchUnitAddressesHash watcher as a filter for ip address
+// changes.
+//
+// It first checks if a cloud service exists for the application, if there is
+// then it returns the net node UUID for the cloud service without checking for
+// the unit's net node since it corresponds to the cloud container address
+// instead.
+//
+// If the unit does not exist an error satisfying
+// [applicationerrors.UnitNotFound] will be returned.
+func (st *State) GetNetNodeUUIDByUnitName(ctx context.Context, name coreunit.Name) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	unitName := unitName{Name: name}
+	k8sServiceNetNodeStmt, err := st.Prepare(`
+SELECT k.net_node_uuid AS &dbUUID.uuid
+FROM   k8s_service k
+JOIN   application a ON a.uuid = k.application_uuid
+JOIN   unit u ON u.application_uuid = a.uuid
+WHERE  u.name = $unitName.name
+`, unitName, dbUUID{})
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	unitNetNodeStmt, err := st.Prepare(`
+SELECT net_node_uuid AS &dbUUID.uuid
+FROM   unit
+WHERE  name = $unitName.name
+`, unitName, dbUUID{})
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	var netNodeUUID dbUUID
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// First try to get the net node UUID from the k8s service.
+		err := tx.Query(ctx, k8sServiceNetNodeStmt, unitName).Get(&netNodeUUID)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Capture(err)
+		} else if err == nil {
+			return nil
+		}
+
+		// If nothing found, try to get the net node UUID from the unit.
+		err = tx.Query(ctx, unitNetNodeStmt, unitName).Get(&netNodeUUID)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Capture(err)
+		} else if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("%w: %s", applicationerrors.UnitNotFound, name)
+		}
+
+		return nil
+	}); err != nil {
+		return "", errors.Capture(err)
+	}
+
+	return netNodeUUID.UUID, nil
+}
+
+// GetAddressesHash returns the sha256 hash of the application unit and cloud
+// service (if any) addresses along with the associated endpoint bindings.
+func (st *State) GetAddressesHash(ctx context.Context, appUUID coreapplication.ID, netNodeUUID string) (string, error) {
```

Comment:
Which changes this name.

Reply:
Answered here https://github.com/juju/juju/pull/19595/files#r2061574159.